File: SwiftOSXProofOfConcept/data2yaml/data/scale.lua
sptbl["scale"] = {

    files = { 
        module = "scale.c",
        header = "scale.h",
        example = "ex_scale.c",
    },
    
    func = {
        create = "sp_scale_create",
        destroy = "sp_scale_destroy",
        init = "sp_scale_init",
        compute = "sp_scale_compute",
    },
    
    params = {
        optional = {
            {
                name = "min",
                type = "SPFLOAT",
                description = "Minimum value to scale to.",
                default = 0
            },
            {
                name = "max",
                type = "SPFLOAT",
                description ="Maximum value to scale to.",
                default = 1
            },
        }
    },
    
    modtype = "module",
    
    description = [[This module scales an incoming signal (in the range of -1, and 1),
and scales it to an arbitrary range.
]], 
    
    ninputs = 1,
    noutputs = 1,
    
    inputs = { 
        {
            name = "in",
            description = "Signal to be scaled, expected to be in the range of +/-1."
        },
    },
    
    outputs = {
        {
            name = "out",
            description = "Signal out"
        },
    }

}

File: AudioKit/Common/Internals/Soundpipe/modules/scale.c
#include <stdlib.h>
#include "soundpipe.h"

int sp_scale_create(sp_scale **p)
{
    *p = malloc(sizeof(sp_scale));
    return SP_OK;
}

int sp_scale_destroy(sp_scale **p)
{
    free(*p);
    return SP_OK;
}

int sp_scale_init(sp_data *sp, sp_scale *p)
{
    p->min = -1;
    p->max = 1;
    return SP_OK;
}

int sp_scale_compute(sp_data *sp, sp_scale *p, SPFLOAT *in, SPFLOAT *out)
{
    *out =  *in * (p->max - p->min) + p->min;
    return SP_OK;
}
